### DISCUSSION TITLE
Merging TimeLapse Images

### DISCUSSION THREAD

**Discussion by Nickcom4** [ImageMagick/ImageMagick#7140] [Consulting]
Using time lapse photography, I'm trying to highlight a specific change through transparent overlaying images. Daily, I need to generate a GIF and a single image file that merges a sequence of 30-75 images. I've used ImageMagick to create a GIF. However, my use of `-evaluate-sequence min` needs some improvement. 

In these images, my ultimate goal is to plot the position of the shadow cast from the screw tip in comparison to a  thin, red-colored horizontal line. Would you please help me with this? I'm trying to show that the shadow tip throughout the day is a curve most of the year. Only 2 days per year is it a straight line.

[Here the example fileset.](https://www.dropbox.com/scl/fo/ndg67g5cjw1kiyt8tw0w1/h?rlkey=0pfurz0chuain5itnrabww0lm&dl=1)

` magick $(ls *.JPG) 20240226_animated.gif` - works well, just need to significantly reduce file size

`magick $(ls *.JPG) -evaluate-sequence min 20240226_Min.jpg` needs improvement:
    1. Brighten the result image so it is very easy to see like the GIF is.
    [optional] 2.Add a red horizontal line that touches the curve plotted by the screw tip's shadow

I'm open to any other suggested improvements in command or my workflow.

**Comment by snibgo** [comment]
I suggest you chop down the tree that causes shadows.

More seriously, I would simplify the images. Crop to the interesting area, convert the gray, threshold to make just black and white, and find all the connected components that are larger than 1000 pixels.
```
set SRC=DSCF0318.JPG

%IMG7%magick ^
  %SRC% ^
  -gravity North ^
  -crop 1286x500+0+125 +repage ^
  -colorspace Gray ^
  -threshold 40%% ^
  -write x0.png ^
  -define connected-components:verbose=true ^
  -define connected-components:mean-color=true ^
  -define connected-components:area-threshold=1000 ^
  -connected-components 4 ^
  x.png
```
This is Windows BAT syntax. I think you are using bash, so change every caret ^ to backslash \\, and every double %% to single %.

The output x.png is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/6c44b371-e5df-4a58-8df4-867bcc88457d)

The text output from the command is:
```
Objects (id: bounding-box centroid area mean-color):
  0: 1286x500+0+0 642.5,250.0 638615 gray(255)
  2: 77x310+617+48 646.1,180.7 4385 gray(0)
```
"gray(0)" is black. So the largest black area has a bounding box width 77 and height 310, with top-left corner at (617,48). So the bottom-right is at x=617+77-1 = 693 and y=48+310-1 = 357.

This is the approximate position of the shadow of the point of the screw. It isn't exact, because the shadow of the thread of the screw is to the right of the point. A search of the local area could pinpoint the exact position.

Repeat this for every image. This gives a list of the coordinates of the shadow of the point of the screw. Draw those as a "polyline", in red or whatever colour you want.

**Comment by snibgo** [comment]
A simpler way to find the point: when scanning the image from top-left across and downwards, it is the last black pixel.
```
magick ^
  %SRC% ^
  -gravity North ^
  -crop 1286x500+0+125 +repage ^
  -colorspace Gray ^
  -threshold 40%% ^
  -define connected-components:verbose=true ^
  -define connected-components:mean-color=true ^
  -define connected-components:area-threshold=1000 ^
  -connected-components 4 ^
  -transparent White ^
  sparse-color:
```
The final entry is:
```
692,357,graya(0,1)
```
... so the point is at (692,357).

**Comment by Nickcom4** [comment]
I'll move the sundial camera system to another location shortly to resolve the tree shadows; thanks for the tip! :-)

This is excellent direction; thank you! What do I owe you and how do I pay?

I'm performing these operations in Terminal on MacOS. 

I'm using `for do done` to batch simplify the images with your instructions. However, I'm having trouble implementing `sparse-color:`

`for f in DSCF*.JPG; do
    echo simplifying $f
    magick $f \
    -gravity North \
    -crop 1286x500+0+125 +repage \
    -colorspace Gray -threshold 40% \
    -define connected-components:verbose=true \
    -define connected-components:mean-color=true \
    -define connected-components:area-threshold=1000 \
    -connected-components 4 \
    -transparent White \
    simple-$f;\
done`

After I get the `sparse-color: ` issue resolved, do you know of a simply way to push all of the coordinates into a polyline with IM?

**Comment by snibgo** [comment]
> What do I owe you and how do I pay?

Nothing.

Writing to `sparse-color:` writes text to stdout. I suggest piping to `tr` to translate spaces to newlines, and piping that to `tail` to retrieve the final line. Like this:
```
magick ^
    ...
    ...
    -connected-components 4 ^
    -transparent White ^
    sparse-color: ^| tr " " "\n" ^| tail -n 1
```
The last line might be:
```
692,357,graya(0,1)
```
You need to strip the trailing ",graya(0,1)", but I forget how to do that in bash.

Append "692,357" to a text file that has been created with a line that contains "polyline". Suppose that file is named `coords.lis`. So that file might start like this:
```
polyline
834,382
816,378
798,375
783,372
```
Then you can draw the that polyline on an image. For example, drawing to a 1000x1000 white image:
```
magick -size 1000x1000 xc: -draw @coords.lis p.png
```
But you could draw on your composite "minimum", or whatever.

**Reply by fmw42** [reply]
This should do what you need using cut

```
var="692,357,graya(0,1)"
new_var=`echo $var | cut -d, -f1 -f2`
echo $new_var
692,357

```

**Comment by Nickcom4** [comment]
This is wonderful; thank you!

**Comment by Nickcom4** [comment]
Thanks to you; image processing is almost automatic now!
![20240226-original-final-opt2](https://github.com/ImageMagick/ImageMagick/assets/6264554/44a946e8-575e-4ddd-a0ed-5b951ed94e43)
